src/serialized_len.rs

```diff
@@ -0,0 +1,128 @@
+use crate::io;
+
+#[derive(Default, Clone, Debug, PartialEq, Eq)]
+pub struct EmptyWrite;
+
+impl io::Write for EmptyWrite {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+#[derive(Default, Clone, Debug, PartialEq, Eq)]
+pub(crate) struct WriteCounterThreshold {
+    counter: usize,
+    threshold: usize,
+}
+
+impl WriteCounterThreshold {
+    pub(crate) fn new(threshold: usize) -> Self {
+        Self { counter: 0, threshold }
+    }
+    pub(crate) fn bytes_written(&self) -> usize {
+        self.counter
+    }
+}
+
+impl io::Write for WriteCounterThreshold {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.counter += buf.len();
+        if self.counter > self.threshold {
+            Err(io::Error::new(io::ErrorKind::Other, "threshold reached"))
```

Comment:
Sadly, this will cause allocation, slowing down the code that was supposed to avoid it. I recommend to just use `io::Error::from(io::ErrorKind::Other)` which doesn't allocate.